### PR TITLE
Extended test coverage for VehicleData tests (internal review)

### PR DESCRIPTION
**Pull request by dboltovskyi**
ATF Test Scripts to check [#FORDTCN-10871](https://adc.luxoft.com/jira/browse/FORDTCN-10871)

This PR is **[ready]** for review.

### Summary
Extended test coverage for Vehicle Data tests

### ATF version
develop

### Changelog

- Introduced new modules
    - `APIHelper` - set of utility functions to work with both APIs: MOBILE and HMI
    - `APITestDataGenerator` - generate values for parameters based on existing API restrictions and desired value type
- Introduced the following test types to verify `GetVD` and `OnVD` RPCs:
    - `VALID_RANDOM_ALL` - generate random valid values for all possible hierarchy of sub-parameters,
    - `VALID_RANDOM` - generate random valid value for each sub-parameter of hierarchy,
    - `INVALID_TYPE` - generate invalid type value for each sub-parameter of hierarchy,
    - `MANDATORY_ONLY` - generate random valid value for only mandatory sub-parameters of hierarchy,
    - `MANDATORY_MISSING` - generate cases when one mandatory sub-parameter is missing,
    - `LOWER_IN_BOUND/UPPER_IN_BOUND` - generate valid values around the boundary for each sub-parameter of hierarchy,
    - `LOWER_OUT_OF_BOUND/UPPER_OUT_OF_BOUND` - generate invalid values around the boundary for each sub-parameter of hierarchy,
    - `ENUM_ITEMS/ BOOL_ITEMS` - generate valid values for each sub-parameter of hierarchy by iterating for all available items (`ENUM`) or true/false (`BOOLEAN`)
    
- Deleted existing VD scripts which are covered by a new tests
- Updated existing scripts
- Updated `vehicle_data.txt` test set

##### Bug Fixes
- Fixed issue of handling API data types without `Common.` prefix
- Improved generation of random values

### TODO:
 - to update `./test_scripts/API/VehicleData/DOP`
 - to update `./test_scripts/API/VehicleData/GpsShiftSupport`

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: test_scripts/API/VehicleData/OnVehicleData/008_OnVD_enum_and_bool_values.lua [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+common.Title("Preconditions")
+common.Step("Clean environment and update preloaded_pt file", common.preconditions)
+common.Step("Start SDL, HMI, connect Mobile, start Session", common.start)
+common.Step("Register App", common.registerApp)
+common.Step("Activate App", common.activateApp)
+
+common.Title("Test")
+common.getTestsForOnVD(testTypes)
+
+common.Title("Postconditions")
```
Comment (GetmanetsIrina) [inline test_scripts/API/VehicleData/OnVehicleData/008_OnVD_enum_and_bool_values.lua:36]: Parameter names that are not checked can be removed.

File: test_scripts/API/VehicleData/OnVehicleData/001_OnVD_Success.lua [hunk 2]
```diff
@@ -16,22 +16,21 @@
...
-for param in common.spairs(common.getVDParams(true)) do
-  common.Title("VD parameter: " .. param)
-  common.Step("RPC " .. common.rpc.sub, common.processSubscriptionRPC, { common.rpc.sub, param })
-  common.Step("RPC " .. common.rpc.on, common.sendOnVehicleData, { param, common.isExpected })
-end
-for param in common.spairs(common.getVDParams(false)) do
-  common.Title("VD parameter: " .. param)
-  common.Step("RPC " .. common.rpc.on, common.sendOnVehicleData, { param, common.isNotExpected })
-end
+common.getTestsForOnVD(testTypes)
```
Comment (GetmanetsIrina) [inline test_scripts/API/VehicleData/OnVehicleData/001_OnVD_Success.lua:33]: Duplicated checks for parent parameters can be removed.

**Review by GetmanetsIrina — CHANGES_REQUESTED** [review]

File: test_scripts/API/VehicleData/common.lua [hunk 3]
```diff
@@ -100,185 +104,88 @@ m.app = {
...
+local boundValueTypeMap = {
+  [m.testType.UPPER_IN_BOUND] = tdg.valueType.UPPER_IN_BOUND,
+  [m.testType.LOWER_IN_BOUND] = tdg.valueType.LOWER_IN_BOUND,
+  [m.testType.UPPER_OUT_OF_BOUND] = tdg.valueType.UPPER_OUT_OF_BOUND,
+  [m.testType.LOWER_OUT_OF_BOUND] = tdg.valueType.LOWER_OUT_OF_BOUND
+}
+
+--[[ Common Functions ]]
+
+--[[ @getAvailableVDParams: Return available for processing VD parameters
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/VehicleData/common.lua:162]: Please notice this function returns only VD parameters without values.

File: test_scripts/API/VehicleData/OnVehicleData/008_OnVD_enum_and_bool_values.lua (hunk 1 above)
Comment (dboltovskyi) [reply test_scripts/API/VehicleData/OnVehicleData/008_OnVD_enum_and_bool_values.lua:36]: This is done intentionally to be sure that some test is not applicable for particular VD parameter.

File: test_scripts/API/VehicleData/GetVehicleData/008_GetVD_app_version_is_less_than_parameter_version.lua [hunk 4]
```diff
@@ -26,19 +26,19 @@ local common = require('test_scripts/API/VehicleData/common')
 local result = "INVALID_DATA"
 
 --[[ Scenario ]]
-for param, version in common.spairs(common.versioningVD) do
-  common.Title("VD parameter: " .. param)
+for _, test in common.spairs(common.getTests(common.rpc.get, common.testType.PARAM_VERSION)) do
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/VehicleData/GetVehicleData/008_GetVD_app_version_is_less_than_parameter_version.lua:29]: Please use latest ATF

File: test_scripts/API/VehicleData/GetVehicleData/007_GetVD_app_version_is_greater_than_parameter_version.lua [hunk 5]
```diff
@@ -23,19 +23,19 @@
 local common = require('test_scripts/API/VehicleData/common')
 
 --[[ Scenario ]]
-for param, version in common.spairs(common.versioningVD) do
-  common.Title("VD parameter: " .. param)
+for _, test in common.spairs(common.getTests(common.rpc.get, common.testType.PARAM_VERSION)) do
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/VehicleData/GetVehicleData/007_GetVD_app_version_is_greater_than_parameter_version.lua:26]: Please use latest ATF

File: user_modules/api/APIHelper.lua [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+m.eventType = {
+  REQUEST = "request",
+  RESPONSE = "response",
+  NOTIFICATION = "notification"
+}
+
+m.dataType = {
+  INTEGER = { type = "Integer", min = -2147483647, max = 2147483647 }, -- min: 2147483648
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply user_modules/api/APIHelper.lua]: Removed in b3d3d3b

File: user_modules/api/APIHelper.lua [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+end
+
+--[[ @getParamsData: Provide hierarchy of parameters and sub-parameters for the defined API function
+--! @parameters:
+--! pAPIType: type of the API, e.g. 'mobile' or 'hmi'
+--! pEventType: type of the event, e.g. 'request', 'response' or 'notification'
+--! pFuncName: name of the API function, e.g. 'GetVehicleData'
+--! @return: structure with hierarchy of parameters
+--]]
+local function getParamsData(pAPI, pEventType, pFunctionName)
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply user_modules/api/APIHelper.lua]: Done in b3d3d3b

File: user_modules/api/APITestDataGenerator.lua [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+  local arrayValueType = m.valueType.VALID_RANDOM
+  if pTypeData.valueTypeArray then arrayValueType = pTypeData.valueTypeArray end
+  local numOfItems = -1
+  if pTypeData.array == true then
+    if arrayValueType == m.valueType.LOWER_IN_BOUND then
+      numOfItems = pTypeData.minsize
+      if not numOfItems or numOfItems == 0 then numOfItems = 1 end
+    elseif arrayValueType == m.valueType.UPPER_IN_BOUND then
+      numOfItems = pTypeData.maxsize
+      if not numOfItems or numOfItems == 0 then numOfItems = 1 end
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply user_modules/api/APITestDataGenerator.lua]: Done in b3d3d3b

File: test_scripts/API/VehicleData/common.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/VehicleData/common.lua]: Done in b3d3d3b

File: test_scripts/API/VehicleData/common.lua [hunk 9]
```diff
@@ -636,4 +490,677 @@ function m.setAppVersion(pParamVersion, pOperator)
...
+--! @parameters: none
+--! @return: table with tests
+--]]
+local function getBoolItemsTests()
+  local tests = {}
+  local dataTypes = { ah.dataType.BOOLEAN.type }
+  local tcs = createTestCases(ah.apiType.HMI, rpcType, m.rpcHMIMap[rpc],
+    m.isMandatory.ALL, m.isArray.ALL, m.isVersion.ALL, dataTypes)
+  for _, tc in pairs(tcs) do
+    for _, item in pairs({true, false}) do
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/VehicleData/common.lua]: Done in b3d3d3b

File: test_scripts/API/VehicleData/common.lua [hunk 10]
```diff
@@ -636,4 +490,677 @@ function m.setAppVersion(pParamVersion, pOperator)
...
+        m.Title("Test type: " .. m.getKeyByValue(m.testType, tt))
+        for _, t in pairs(tests) do
+          m.Step(t.name, m.processRequest, { t.params })
+        end
+      end
+    end
+  end
+end
+
+--[[ @getTestsForGetVD: Generate test steps for 'OnVehicleData' tests for defined test types
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/VehicleData/common.lua]: Done in b3d3d3b

File: test_scripts/API/VehicleData/GetVehicleData/009_GetVD_min_max_boundary_values.lua [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+-- SDL does:
+-- - a) send GetVehicleData response with (success = true, resultCode = "SUCCESS",
+--    <vd_param> = <data received from HMI>) to App
+-- 3) App sends valid GetVehicleData(<vd_param>=true) request to SDL
+-- SDL does:
+-- - a) transfer this request to HMI
+-- 4) HMI sends VI.GetVehicleData response with invalid data for <vd_param> to SDL
+-- (closest lower/upper values to the defined boundary)
+-- SDL does:
+-- - a) send GetVehicleData response with (success = false, resultCode = "GENERIC_ERROR"
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/VehicleData/GetVehicleData/009_GetVD_min_max_boundary_values.lua]: Done in b3d3d3b

File: user_modules/api/APITestDataGenerator.lua [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+--! @return: value for the parameter
+--]]
+local function getBooleanValue(pTypeData)
+  if pTypeData.data and #pTypeData.data == 1 then
+    return pTypeData.data[1]
+  end
+  if pTypeData.valueType == m.valueType.VALID_RANDOM then
+    return math.random(0, 1) == 1
+  elseif pTypeData.valueType == m.valueType.INVALID_TYPE then
+    return 123
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply user_modules/api/APITestDataGenerator.lua:167]: Done in b3d3d3b

File: test_scripts/API/VehicleData/OnVehicleData/001_OnVD_Success.lua (hunk 2 above)
Comment (dboltovskyi) [reply test_scripts/API/VehicleData/OnVehicleData/001_OnVD_Success.lua:34]: Done in b3d3d3b

File: test_scripts/API/VehicleData/GetVehicleData/009_GetVD_min_max_boundary_values.lua [hunk 13]
```diff
@@ -0,0 +1,48 @@
+----------------------------------------------------------------------------------------------------
+-- Description: Check that SDL processes GetVehicleData RPC with <vd_param> parameter
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/VehicleData/GetVehicleData/009_GetVD_min_max_boundary_values.lua:2]: Done in 48c6278

**Comment by dboltovskyi** [comment]
@GetmanetsIrina Please notice all comments have been addressed in b3d3d3b and 48c6278

File: user_modules/api/APIHelper.lua [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+end
+
+--[[ @getParamsData: Provide hierarchy of parameters and sub-parameters for the defined API function
+--! @parameters:
+--! pAPIType: type of the API, e.g. 'mobile' or 'hmi'
+--! pEventType: type of the event, e.g. 'request', 'response' or 'notification'
+--! pFuncName: name of the API function, e.g. 'GetVehicleData'
+--! @return: structure with hierarchy of parameters
+--]]
+local function getParamsData(pAPI, pEventType, pFuncName)
```
Comment (GetmanetsIrina) [inline user_modules/api/APIHelper.lua:65]:
```suggestion
local function getParamsData(pAPIType, pEventType, pFuncName)
```

**Review by GetmanetsIrina — APPROVED** [review]

File: user_modules/api/APIHelper.lua (hunk 14 above)
Comment (dboltovskyi) [reply user_modules/api/APIHelper.lua]: Done in 1964bea

**Pull request merged**